pages/api/create-member.tsx

```diff
@@ -2,66 +2,132 @@ import {
 import SendGrid from "@sendgrid/mail";
-import airtable from "airtable";
-
-airtable.configure({
-  endpointUrl: "https://api.airtable.com",
-  apiKey: process.env.AIRTABLE_KEY,
-});
+import {
+  addDoc,
+  arrayUnion,
+  collection,
+  doc,
+  DocumentReference,
+  getDocs,
+  query,
+  serverTimestamp,
+  updateDoc,
+  where,
+} from "firebase/firestore";
+import { useEmailCloaker } from "helpers";
 
 SendGrid.setApiKey(process.env.SENDGRID_API_KEY);
 Client.setApiKey(process.env.SENDGRID_API_KEY);
 
-export interface RecordFields {
-  name: string;
-  table: string;
-}
-
-const findRecord = async ({
-  name,
-  table,
-}: RecordFields): Promise<string | boolean> => {
-  return new Promise((resolve, reject) => {
-    airtable
-      .base(process.env.AIRTABLE_BASE)(table)
-      .select({
-        view: "All",
-        filterByFormula: `{Name} = "${name}"`,
-      })
-      .firstPage((error, records) => {
-        if (error) reject(error);
-        resolve(records?.length >= 1 ? records[0].getId() : false);
-      });
+const addPendingReviewRecord = async (
+  docReviewRef: DocumentReference,
+  collectionName: string
+) => {
+  const collectionRef = collection(db, "review");
+  const docRef = doc(collectionRef, collectionName);
+  await updateDoc(docRef, {
+    [collectionName]: arrayUnion(docReviewRef),
+    last_modified: serverTimestamp(),
   });
 };
 
-const addPendingRecord = async ({
-  name,
-  table,
-}: RecordFields): Promise<string> => {
-  let focus = {
-    Name: name,
-    Status: "Pending",
+const addLabelRef = async (
+  label: string,
+  collectionName: string
+): Promise<DocumentReference> => {
+  const collectionRef = collection(db, collectionName);
+  const q = query(collectionRef, where("name", "==", label));
+  const querySnapshot = await getDocs(q);
+  let docRef: DocumentReference;
+  if (!querySnapshot.empty) {
+    // Catches the case where the label already exists but it's pending review
+    docRef = querySnapshot.docs[0].ref;
+  } else {
+    docRef = await addDoc(collectionRef, {
+      name: label,
+      status: StatusEnum.PENDING,
+      last_modified: serverTimestamp(),
+      members: [],
+    });
+  }
+  addPendingReviewRecord(docRef, collectionName);
+  return docRef;
+};
+
+const addMemberToLabels = async (
+  labelReferences: DocumentReference[],
+  memberRef: DocumentReference
+) => {
+  for (const labelRef of labelReferences) {
+    await updateDoc(labelRef, {
+      members: arrayUnion(memberRef),
+      last_modified: serverTimestamp(),
+    });
+  }
+};
+
+const addMember = async (member: MemberFields): Promise<DocumentReference> => {
+  const collectionRef = collection(db, "members");
+  const masked_email = useEmailCloaker(member.email);
+  const masked_email_str = `${masked_email[0]}...${masked_email[1]}${masked_email[2]}`;
+  const data = {
+    ...member,
+    last_modified: serverTimestamp(),
+    last_modified_by: "",
+    masked_email: masked_email_str,
```

Comment:
Do we need to store the masked email? Can we just cloak it when we show the data?

Reply:
had some implementation ideas for this, can bring this convo in our group chat with @taylorkmho

Reply:
Could that lead to the email being exposed when we return the user's data in a response? I'm recalling there are cases where you could have (e.g. the edit form) where we show it, but don't require any authorization.

Reply:
The emails are cloaked in the [emailLookup function](https://github.com/hawaiians/hawaiiansintech/blob/696404be9579e340b19a59da263bfafcf78d0ac7/lib/api.ts#LL124C15-L124C26) and done server side or via API

It's still done here, but we're _also_ saving the masked email (cloaked email) as well. This implementation is secure so long as our keys are safe. I'm just thinking it might be better practice and safer to migrate the emails to another table that's admin-only access for additional security